File: week_1/percolation/weighted_quick_union_uf.rb
class WeightedQuickUnionUF

attr_reader :array

  def initialize n
    @n = n-1
    @array = (0..@n).to_a
    @size = {}
    @array.each do |k|
      @size[k] = 0
    end
  end

  def connected x, y
    root(x) == root(y)
  end

  def union x, y
    i, j = root(x), root(y)
    return if i == j
    if @size[i] < @size[j]
      @array[i] = j
      @size[j] += @size[i]
    else
      @array[j] = i
      @size[i] += @size[j]
    end
  end

  private

  def root (i)
    while i != @array[i] do
      @array[i] = @array[@array[i]]
      i = @array[i]
    end
    return i
  end

end

File: week_1/percolation/weighted_quick_union_uf.py
class WeightedQuickUnionUF:
    
    def __init__(self, n):
      self.n = n
      self.array = list(range(n))
      self.size = { key: 0 for key in self.array }

    def connected(self, x, y):
      return self.__root(x) == self.__root(y)

    def union(self, x, y):
      i, j = self.__root(x), self.__root(y)
      if i == j:
        return
      elif self.size[i] < self.size[j]:
        self.array[i] = j
        self.size[j] += self.size[i]
      else:
        self.array[j] = i
        self.size[i] += self.size[j]

    def __root(self, i):
      while i != self.array[i]:
        self.array[i] = self.array[self.array[i]]
        i = self.array[i]
      return i